File: bots/python/binance_client.py
"""
Binance Pay API Client
Author: Md Rahul Islam <https://github.com/mdrahulislammri>
"""

import time
import hmac
import hashlib
import requests
from typing import Optional


class BinanceClient:
    BASE_URL = "https://api.binance.com"

    def __init__(self, api_key: str, api_secret: str):
        self.api_key    = api_key
        self.api_secret = api_secret
        self.last_error = ""

    def verify_by_order_id(self, order_id: str, amount: float, coin: str = "USDT", max_age_hours: int = 24) -> Optional[dict]:
        transactions = self.get_pay_transactions(100)
        if transactions is None:
            return None

        for tx in transactions:
            if tx.get("orderId") != order_id:
                continue

            # Incoming = positive amount
            tx_amount = float(tx.get("amount", 0))
            if tx_amount <= 0:
                self.last_error = "This is an outgoing payment, not an incoming one."
                return None

            # Coin check
            if tx.get("currency", "").upper() != coin.upper():
                self.last_error = f"Coin mismatch. Expected {coin} but got {tx.get('currency', '')}."
                return None

            # Amount check
            if abs(tx_amount - amount) > 0.01:
                self.last_error = f"Amount mismatch. Expected {amount:.2f} but got {tx_amount:.2f} {coin}."
                return None

            # Age check — transactionTime is in milliseconds
            tx_time = int(tx.get("transactionTime", 0)) / 1000
            if tx_time > 0 and (time.time() - tx_time) > (max_age_hours * 3600):
                self.last_error = f"Transaction is older than {max_age_hours} hours."
                return None

            return tx

        self.last_error = "Order ID not found in Binance Pay history."
        return None

    def get_pay_transactions(self, limit: int = 100) -> Optional[list]:
        params    = {"limit": min(limit, 100), "timestamp": int(time.time() * 1000)}
        query     = "&".join(f"{k}={v}" for k, v in params.items())
        signature = hmac.new(self.api_secret.encode(), query.encode(), hashlib.sha256).hexdigest()
        url       = f"{self.BASE_URL}/sapi/v1/pay/transactions?{query}&signature={signature}"

        try:
            resp = requests.get(url, headers={"X-MBX-APIKEY": self.api_key}, timeout=30)
            data = resp.json()
        except Exception as e:
            self.last_error = f"Connection error: {e}"
            return None

        if data.get("code") != "000000":
            self.last_error = data.get("errorMessage") or data.get("message") or "Failed to fetch transactions."
            return None

        return data.get("data", [])

    def get_error(self) -> str:
        return self.last_error or "Unknown error."

File: scripts/web/BinanceClient.php
<?php
/**
 * Binance Pay Client — Standalone Web Scripts
 * @author Md Rahul Islam <https://github.com/mdrahulislammri>
 */

class BinanceClient
{
    const BASE_URL = 'https://api.binance.com';

    private string $apiKey;
    private string $apiSecret;
    private string $lastError = '';

    public function __construct(string $apiKey, string $apiSecret)
    {
        $this->apiKey    = $apiKey;
        $this->apiSecret = $apiSecret;
    }

    public function verifyByOrderId(string $orderId, float $amount, string $coin = 'USDT', int $maxAgeHours = 24): ?array
    {
        $transactions = $this->getPayTransactions(100);
        if ($transactions === null) return null;

        foreach ($transactions as $tx) {
            if (($tx['orderId'] ?? '') !== $orderId) continue;

            $txAmount = (float)($tx['amount'] ?? 0);
            if ($txAmount <= 0) { $this->lastError = 'This is an outgoing payment.'; return null; }
            if (strtoupper($tx['currency'] ?? '') !== strtoupper($coin)) { $this->lastError = 'Coin mismatch.'; return null; }
            if (abs($txAmount - $amount) > 0.01) { $this->lastError = sprintf('Amount mismatch. Expected %.2f but got %.2f.', $amount, $txAmount); return null; }

            $txTime = (int)(($tx['transactionTime'] ?? 0) / 1000);
            if ($txTime > 0 && (time() - $txTime) > ($maxAgeHours * 3600)) { $this->lastError = 'Transaction too old.'; return null; }

            return $tx;
        }

        $this->lastError = 'Order ID not found.';
        return null;
    }

    public function getPayTransactions(int $limit = 100): ?array
    {
        $response = $this->signedGet('/sapi/v1/pay/transactions', ['limit' => min($limit, 100)]);
        if ($response === null) return null;
        if (($response['code'] ?? '') !== '000000') { $this->lastError = $response['message'] ?? 'API error.'; return null; }
        return $response['data'] ?? [];
    }

    public function getDepositHistory(string $coin = 'USDT', int $limit = 100): ?array
    {
        $response = $this->signedGet('/sapi/v1/capital/deposit/hisrec', ['coin' => strtoupper($coin), 'limit' => $limit, 'recvWindow' => 60000]);
        if ($response === null) return null;
        if (!is_array($response)) { $this->lastError = 'Unexpected response.'; return null; }
        return $response;
    }

    public function getError(): string { return $this->lastError ?: 'Unknown error.'; }

    private function signedGet(string $path, array $params = []): ?array
    {
        $params['timestamp'] = (int)(microtime(true) * 1000);
        $query     = http_build_query($params);
        $signature = hash_hmac('sha256', $query, $this->apiSecret);
        $url       = self::BASE_URL . $path . '?' . $query . '&signature=' . $signature;

        $ch = curl_init($url);
        curl_setopt_array($ch, [CURLOPT_RETURNTRANSFER => true, CURLOPT_HTTPHEADER => ['X-MBX-APIKEY: ' . $this->apiKey], CURLOPT_TIMEOUT => 30, CURLOPT_SSL_VERIFYPEER => true]);
        $raw = curl_exec($ch);
        $err = curl_error($ch);
        curl_close($ch);

        if ($err || !$raw) { $this->lastError = 'Connection error: ' . $err; return null; }
        $decoded = json_decode($raw, true);
        if (json_last_error() !== JSON_ERROR_NONE) { $this->lastError = 'Invalid response.'; return null; }
        if (isset($decoded['code']) && is_int($decoded['code']) && $decoded['code'] < 0) { $this->lastError = $decoded['msg'] ?? 'API error.'; return null; }
        return $decoded;
    }
}
